Add support for static private IP addressing and DNS resolution for external slurmdbd instance

jdeamicis opened the pull request:
### Description of changes
* Add mechanism to assign a static secondary private IP to the external slurmdbd instance (via `user_data.sh`).
* Add Route53 hosted zone to the external slurmdbd stack
* Add DNS record for the external slurmdbd instance (only for the case of `CfnInstance`, not via ASG).

### Tests
* Manually provisioned the stack successfully.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Set up private IP for external slurmdbd instance

Add components to allow the assignment of static private secondary
IP address to the slurmdbd instance.

Add possibility to set up the external slurmdbd instance as a
standalone instance in the CDK app rather than using an ASG.

Minor improvements to the provisional user_data.sh file.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add Route53 DNS to external Slurmdbd stack

Add hosted zone belonging to the stack.

Add DNS record related to instance being generated by the
stack (only in the case of CfnInstance, not via the Auto
Scaling Group).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cloudformation/external-slurmdbd/external_slurmdbd/external_slurmdbd_stack.py:
```diff
@@ -361,3 +420,29 @@ def _add_cloudwatch_log_group(self):
+    def _add_hosted_zone(self):
+        return route53.CfnHostedZone(
+            self,
+            id="ExternalSlurmdbdHostedZone",
```
Shouldn't this be provided as parameter ?

jdeamicis replied:
I will create an internal task to further dive deep on this.

NSsirena commented on cloudformation/external-slurmdbd/resources/user_data.sh:
```diff
@@ -7,13 +7,22 @@ function vendor_cookbook
+ENI_ID=$(aws ec2 describe-instances --instance-ids ${!INSTANCE_ID} --region ${Region} | jq .Reservations[0].Instances[0].NetworkInterfaces[0].NetworkInterfaceId | tr -d '"')
+
+aws ec2 assign-private-ip-addresses --region ${Region} --network-interface-id ${!ENI_ID} --private-ip-addresses ${PrivateIp} --allow-reassignment
+
```
Will failures be logged or do we need some extra check to detect failures and log them ?

jdeamicis replied:
At the moment they would just be logged in the cloud-init output. I will create an internal task to make sure we address this.

EddyMM commented on cloudformation/external-slurmdbd/resources/user_data.sh:
```diff
@@ -7,13 +7,22 @@ function vendor_cookbook
+
+aws ec2 assign-private-ip-addresses --region ${Region} --network-interface-id ${!ENI_ID} --private-ip-addresses ${PrivateIp} --allow-reassignment
+
+ip addr add ${PrivateIp}/${SubnetPrefix} dev eth0
```
Can we try to test the assignment after associating the PrivateIP to the device (a `ping` maybe)?

This is to avoid proceeding with vendoring the cookbook yet the IP address wasn't associated as expected. This would help rule out some causes of slurmctld <---> slurmdbd connection problems if they occur later.

jdeamicis replied:
Let me address this quickly, this should be possible within the user-data script.

jdeamicis replied:
Implemented.

jdeamicis committed:
Add check for private IP assignment

Add bash function to check that the secondary private IP was
correctly assigned to the instance via EC2.

Add DescribeNetworkInterfaces policy to allow the implemented
check.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
